### DISCUSSION TITLE
Next.js Server Actions

### DISCUSSION THREAD

**Discussion by leaen** [amannn/next-intl#353] [Q&A]
I am using Next.js server actions for updating my database. But using next-intl middleware it throws below error in console: 

POST http://localhost:3000/account 405 (Method Not Allowed) server-action-reducer.js:23 
   
Thank you in advance for any help!

**Comment by irg1008** [comment]
Were you able to solve this?

**Comment by irg1008** [comment]
Hi I think we should address this.
Personally in my app I make all database and third party api calls inside server actions.
Not being able to add i18n while keeping the functions is a bit a pain in the ass.

Seeing that they are probably very soon becoming beta feature and soon enough stable, is there any plan to make it work, or do we now if it's a bug / something we can look for in the i18n middleware?

**Reply by amannn** [reply]
I haven't experimented much with Server Actions yet, but would certainly be interested in learning more how `next-intl` could integrate with them.

Can you open a bug report that includes details about what you're trying to achieve and which errors you're encountering? The feature is currently alpha as far as I know, so also links to potentially related issues on the Next.js issue tracker would help. Sometimes there's work necessary on the Next.js side, but it would be helpful to find out if this is the case, or if `next-intl` can adapt somewhere.

Thanks!

**Comment by irg1008** [comment]
They have an issue opened at Nex.js: https://github.com/vercel/next.js/issues/54992
It seems to be a problem related with generateStaticParams

**Comment by irg1008** [comment]
Guys the vercel team fixed the issue: https://github.com/vercel/next.js/issues/49408#issuecomment-1720291375

They are adding this to Next.js 13.5 next week!!

**Comment by irg1008** [comment]
It's already available on 13.4.20-canary.31

When downloading this version it works but I noticed something.

Redirect calls revalidate the path more than one time (n times), being n the number of locales in our project. In my case I have 3 locales so when I call redirect (next-intl or next.js, happens with both), the layout revalidation (and therefore server action call) happens three times

I think the problem has to do with `getStaticParams` function, but I don't know how to solve it

Enabling just one locale works as it should, or using the defaultLocale only (the log comes from the server action that checks logged status after log in):

Using defaultLocale after log in:
![image](https://github.com/amannn/next-intl/assets/47559427/3d8f90b1-a917-4caa-9a05-04c9aac80a06)

Using any non-default locale:
![image](https://github.com/amannn/next-intl/assets/47559427/d0ff4607-ec2d-4062-9578-c96de2503767)

Here we see the server actions is called twice, this prevents handling manual redirects and prevents forms submits

**Comment by alexleach** [comment]
I was having the same 405 error message in the Developer Tools Console, and managed to follow @AndrewRPorter's advice in the [next.js Issue](https://github.com/vercel/next.js/issues/49408#issuecomment-1703337599) of splitting the layout into two files, one at `app/layout.tsx`, and another at `app/[locale]/layout.tsx`. No working example was provided, but I've managed to get it work, so thought I'd share my working layouts... I'm using Next.js 14.1.0 and next-intl 3.8.0.

```typescript
// app/layout.tsx
/**
 * The Layout in this file deals with setting the locale and translations for
 * the page. A nested layout then deals with the rest of the app layout.
 */
import * as React from 'react';

import { getTranslations } from 'next-intl/server';
import { unstable_setRequestLocale } from 'next-intl/server';
import { locales } from '@/i18n';

export async function generateMetadata({params: {locale}}: { params: {locale: string}} ) {
  const t = await getTranslations({locale, namespace: 'Metadata'});

  return {
    title: t('title'),
    description: t('description')
  };
}

export function generateStaticParams() {
  return locales.map((locale) => ({locale}));
}

export default async function Layout({ children, params: {locale}} : { children: React.ReactNode, params: {locale: string} }) {
  unstable_setRequestLocale(locale);

  return (
    <html lang={locale}>
      <body>
        {children}
      </body>
    </html>
  );
}

```

```typescript
// app/[locale]/layout.tsx
import * as React from 'react';

import LoadingIcon from '@/app/loading';

import Header from '@/components/Header';
import Footer from '@/components/Footer'

export default async function LocaleLayout({ children } : { children: React.ReactNode }) {

  return (
    <Header />
    <main>
      <React.Suspense fallback={<LoadingIcon />} >
        {children}
      </React.Suspense>
    </main>
    <Footer />
  );
}
```

I also seem to have some issues with the default `matcher` configuration (commented out below), which for some reason appears to match all paths.

<img width="1010" alt="Error with default matcher configuration" src="https://github.com/amannn/next-intl/assets/375981/5eb54267-5e47-4ae0-b1d0-98ab06ff95a2">

This is I think unrelated to the above issues, but either way, instead I have had to use the configuration below.

```typescript
// middleware.ts

import createMiddleware from 'next-intl/middleware';

export default createMiddleware({
  // A list of all locales that are supported
  locales: ['en', 'fr', 'de'],
  // Used when no locale matches
  defaultLocale: 'en',
  // Don't add a locale prefix
  localePrefix: 'as-needed'
});

export const config = {
  matcher: [
    '/((?!api|_next|_vercel|__session|.*\\..*).*)',
    // Using only the below paths seems to match every path...
    // '/',
    // '/(en|fr|de)/:path*'
  ]
};
```


Hopefully that's a nice minimal example that might help someone else. FWIW, I'm using [`firebaseui`](https://github.com/firebase/firebaseui-web), and the 405 message occurred when trying to access `https://localhost:5002/__session` on my dev machine.

**Reply by globally-locals** [reply]
have you solved your issue?

**Comment by thereis** [comment]
Well, the issue seems to not be fixed! I am facing this problem when I call a server action from any localized subpath, for example, `/profile`.

It's working in the local development, but for any preview/prod it throws a `405 (Method Not Allowed)`!

**Reply by thereis** [reply]
The issue is pretty straight forward, if you set the page to be `force-dynamic`, all the server actions will be inutile.

You will not face this issue on development environment because Next doesn't generate the static content in your local environment.

Beware.
